File: examplesForThesis-R/ch3-example-forgettingproperties.R
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# Mixing property in the LGSS model
# Example 3.5 in Section 3.4.1
#
########################################################################

stateseqs1  <- read.table("ch3-example-forgettingproperties1.csv",header=T,sep=",",stringsAsFactors=F)[,-1]
stateseqs2  <- read.table("ch3-example-forgettingproperties2.csv",header=T,sep=",",stringsAsFactors=F)[,-1]
stateseqs3  <- read.table("ch3-example-forgettingproperties3.csv",header=T,sep=",",stringsAsFactors=F)[,-1]

T = 21;
N = 10;

layout(matrix(1:3, 3, 1, byrow = TRUE))  
par(mar=c(4,5,1,1)) 

plot(0:14,stateseqs1[,1],xlim=c(0,8),ylim=c(-60,60),xlab="Time",ylab="State",cex.axis=0.95,type="l",lwd=2,col="steelblue")

for (ii in 2:20) {
  lines(0:14,stateseqs1[,ii],lwd=2,col="steelblue")
}
text(8.2,55,expression(phi * "=0.2"),col="steelblue",font=2,pos=2,cex=1.4)

plot(0:14,stateseqs2[,1],xlim=c(0,8),ylim=c(-60,60),xlab="Time",ylab="State",cex.axis=0.95,type="l",lwd=2,col="darkred")

for (ii in 2:20) {
  lines(0:14,stateseqs2[,ii],lwd=2,col="darkred")
}
text(8.2,55,expression(phi * "=0.5"),col="darkred",font=2,pos=2,cex=1.4)

plot(0:14,stateseqs3[,1],xlim=c(0,8),ylim=c(-60,60),xlab="Time",ylab="State",cex.axis=0.95,type="l",lwd=2,col="darkgreen")

for (ii in 2:20) {
  lines(0:14,stateseqs3[,ii],lwd=2,col="darkgreen")
}
text(8.2,55,expression(phi * "=0.8"),col="darkgreen",font=2,pos=2,cex=1.4)

########################################################################
# End of file
########################################################################

File: examplesForThesis-python/ch3-example-forgettingproperties.py
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# Mixing property in the LGSS model
# Example 3.5 in Section 3.4.1
#
########################################################################

from kf import *
from pmh import *
from classes import *
from helpers import *
import pandas
import numpy as np

########################################################################
# Arrange the data structures
########################################################################
data             = stData();
kf               = kalmanFilter();
par              = stParameters();
mh               = stPMH();

########################################################################
# Setup the system
########################################################################
sys              = stSystemLGSS()
sys.version      = "standard"
sys.par          = np.zeros((3,1))
sys.par[0]       = 0.50;
sys.par[1]       = 1.00;
sys.par[2]       = 0.10;
sys.T            = 15;

thSys            = stSystemLGSS()
thSys.T          = sys.T;
thSys.version    = sys.version

########################################################################
# Setup the parameters for the algorithm
########################################################################

par.fileprefix   = "lgss"
par.nPars          = 1;
par.normLimit      = 0.1;
par.verboseSampler = 0;

par.Po           = 0.01;
par.xo           = 0;

########################################################################
# Run the sampler
########################################################################

nRuns = 20;
xx = np.zeros((nRuns,sys.T));

# Mixing at phi=0.20
sys.par[0]       = 0.20;
for ii in range(0,nRuns):
    
    # Generate initial state
    sys.xo = np.random.normal(0,20);
    
    # Generate data and save the state
    data.sample(sys,np.zeros(sys.T))
    xx[ii,:] = (data.x).transpose();

# Export data to R for plotting
plot(xx.transpose())
pandas.DataFrame(xx.transpose()).to_csv("ch3-example-forgettingproperties1.csv");

sys.par[0]       = 0.50;
for ii in range(0,nRuns):
    
    # Generate initial state
    sys.xo = np.random.normal(0,20);
    
    # Generate data and save the state
    data.sample(sys,np.zeros(sys.T))
    xx[ii,:] = (data.x).transpose();

# Export data to R for plotting
plot(xx.transpose())
pandas.DataFrame(xx.transpose()).to_csv("ch3-example-forgettingproperties2.csv");

sys.par[0]       = 0.80;
for ii in range(0,nRuns):
    
    # Generate initial state
    sys.xo = np.random.normal(0,20);
    
    # Generate data and save the state
    data.sample(sys,np.zeros(sys.T))
    xx[ii,:] = (data.x).transpose();

# Export data to R for plotting
plot(xx.transpose())
pandas.DataFrame(xx.transpose()).to_csv("ch3-example-forgettingproperties3.csv");

########################################################################
# End of file
########################################################################
